In [1]:
import pandas as pd
import numpy as np
from sklearn.neural_network import MLPClassifier
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.multioutput import MultiOutputClassifier
from sklearn.metrics import classification_report, accuracy_score
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

# CARGA DE DATOS 
train_data = pd.read_csv("../../Data/Chi2/train_2000_chi2.csv")
test_data = pd.read_csv("../../Data/test_indexado.csv")

# LLenar los registros vacíos del train
train_data['Text'] = train_data['Text'].fillna('')

# Definir las clases de emociones
emotion_classes = train_data.columns[2:].tolist()

In [2]:
# TF-IDF VECTORIZACIÓN
vectorizer = TfidfVectorizer(lowercase=True, stop_words="english", strip_accents="unicode", max_features=5000)
X_train = vectorizer.fit_transform(train_data['Text'].values)
X_test = vectorizer.transform(test_data['Text'].values)
y_train = np.asarray(train_data[emotion_classes])
y_test = np.asarray(test_data[emotion_classes])

In [3]:
#Spliteo del conjunto de entrenamiento para validación
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)


# Multilayer Neural Network Classifier para clasificación multi-etiqueta
base_model = MLPClassifier(max_iter=1000)
mlp_model = MultiOutputClassifier(base_model)
mlp_model.fit(X_train, y_train)

#Predicción
y_pred = mlp_model.predict(X_val)

In [ ]:
# EVALUACIÓN

print("\nResultados Neural Network Classifier:\n")
print("Accuracy:", accuracy_score(y_val, y_pred))
print("\nReporte de clasificación:\n", classification_report(y_val, y_pred, target_names=emotion_classes, zero_division=0))



Resultados Neural Network Classifier:

Accuracy: 0.3671964985026492

Reporte de clasificación:
                 precision    recall  f1-score   support

    admiration       0.66      0.50      0.57       863
     amusement       0.73      0.68      0.70       453
         anger       0.48      0.24      0.33       323
     annoyance       0.35      0.15      0.21       483
      approval       0.32      0.12      0.17       577
        caring       0.32      0.15      0.21       212
     confusion       0.31      0.13      0.18       258
     curiosity       0.40      0.10      0.16       460
        desire       0.45      0.25      0.32       128
disappointment       0.33      0.10      0.15       244
   disapproval       0.25      0.10      0.14       383
       disgust       0.47      0.30      0.37       156
 embarrassment       0.67      0.31      0.42        58
    excitement       0.42      0.18      0.25       175
          fear       0.54      0.32      0.40       116
     g


Tardó 95 minutos en entrenar el modelo.

Resultados Neural Network Classifier:

Accuracy: 0.3671964985026492

Reporte de clasificación:
                 precision    recall  f1-score   support

    admiration       0.66      0.50      0.57       863
     amusement       0.73      0.68      0.70       453
         anger       0.48      0.24      0.33       323
     annoyance       0.35      0.15      0.21       483
      approval       0.32      0.12      0.17       577
        caring       0.32      0.15      0.21       212
     confusion       0.31      0.13      0.18       258
     curiosity       0.40      0.10      0.16       460
        desire       0.45      0.25      0.32       128
disappointment       0.33      0.10      0.15       244
   disapproval       0.25      0.10      0.14       383
       disgust       0.47      0.30      0.37       156
 embarrassment       0.67      0.31      0.42        58
    excitement       0.42      0.18      0.25       175
          fear       0.54      0.32      0.40       116
     gratitude       0.91      0.84      0.88       544
         grief       0.00      0.00      0.00        14
...
     macro avg       0.46      0.29      0.34     10234
  weighted avg       0.53      0.42      0.45     10234
   samples avg       0.45      0.44      0.44     1023